statistics/row_sampler.go

```diff
@@ -163,35 +176,49 @@ func (s *RowSampleBuilder) Collect() (*RowSampleCollector, error) {
 	}
 }
 
-func (s *RowSampleCollector) collectColumns(sc *stmtctx.StatementContext, cols []types.Datum) error {
+func (s *RowSampleCollector) collectColumns(sc *stmtctx.StatementContext, cols []*types.Datum) error {
 	for i, col := range cols {
 		if col.IsNull() {
 			s.NullCount[i]++
 			continue
 		}
 		s.TotalSizes[i] += int64(len(col.GetBytes())) - 1
 		// Minus one is to remove the flag byte.
-		err := s.FMSketches[i].InsertValue(sc, col)
+		err := s.FMSketches[i].InsertValue(sc, *col)
 		if err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func (s *RowSampleCollector) collectColumnGroups(sc *stmtctx.StatementContext, cols []types.Datum, colGroups [][]int64) error {
+func (s *RowSampleCollector) collectColumnGroups(sc *stmtctx.StatementContext, cols, originalColDatum []*types.Datum, collator []collate.Collator, colGroups []*tipb.AnalyzeColumnGroup) error {
 	colLen := len(cols)
 	datumBuffer := make([]types.Datum, 0, len(cols))
+	var tmpDatum types.Datum
 	for i, group := range colGroups {
 		datumBuffer = datumBuffer[:0]
 		hasNull := true
-		for _, c := range group {
-			datumBuffer = append(datumBuffer, cols[c])
+		for ci, c := range group.ColumnOffsets {
+			if group.PrefixLengths[ci] != types.UnspecifiedLength && collator[c] != nil && !cols[c].IsNull() {
+				originalColDatum[c].Copy(&tmpDatum)
+				s.cutDatumByPrefixLength(&tmpDatum, int(group.PrefixLengths[i]), collator[c])
```

Comment:
How about directly using `CutDatumByPrefixLen()` here since we can get the `FieldTypes` in `Collect()`?
And I think it's a little risky to keep the consistensy between `CutDatumByPrefixLen()` and `(*RowSampleCollector).cutDatumByPrefixLength()` if there are changes in the future.